tools/storage/uri/src/test/java/org/projectnessie/storage/uri/TestStorageUri.java

```diff
@@ -0,0 +1,212 @@
+package org.projectnessie.storage.uri;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class TestStorageUri {
+
+  private static URI normalizedURI(String value) {
+    URI u = URI.create(value).normalize();
+    try {
+      if (u.isOpaque()) {
+        return new URI(u.getScheme(), u.getSchemeSpecificPart(), u.getFragment());
+      } else {
+        return new URI(
+            u.getScheme(),
+            u.getUserInfo(),
+            u.getHost(),
+            u.getPort(),
+            u.getPath(),
+            u.getQuery(),
+            u.getFragment());
+      }
+    } catch (URISyntaxException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Test
+  void testScheme() {
+    assertThat(StorageUri.of("/path/to/file").scheme()).isNull();
+    assertThat(StorageUri.of("relative/to/file").scheme()).isNull();
+    assertThat(StorageUri.of("file:/path/to/file").scheme()).isEqualTo("file");
+    assertThat(StorageUri.of(URI.create("file:/path/to/file")).scheme()).isEqualTo("file");
+  }
+
+  @ParameterizedTest
+  @CsvSource({
+    "/absolute/path,/absolute/path",
+    "/absolute/path,/absolute/path/",
+    "//extra/leading,//extra/leading",
+    "//extra/leading,//extra/leading/",
+    "relative/path,relative/path",
+    "relative/path,relative/path/",
+    "file:/path,file:///path",
+    "file:///path,file:///path/",
+    "file://////path,file:///path",
+    "file:/path,/path",
+    "file:/path,path",
+    "s3://b/,s3://b",
+    "s3://b/,s3://b/",
+    "s3://b/path,s3://b/path",
+    "s3://b/path,s3://b/path/",
+    "s3://b/path,s3://b/path//",
+    "s3://b/path,/path",
+  })
+  void testCompare(String v1, String v2) {
+    StorageUri s1 = StorageUri.of(v1);
+    StorageUri s2 = StorageUri.of(v2);
+    URI u1 = normalizedURI(v1);
+    URI u2 = normalizedURI(v2);
+    assertThat(s1.equals(s2)).isEqualTo(u1.equals(u2));
+    assertThat(s2.equals(s1)).isEqualTo(u2.equals(u1));
+    assertThat(Integer.signum(s2.compareTo(s1))).isEqualTo(Integer.signum(u2.compareTo(u1)));
+    assertThat(Integer.signum(s1.compareTo(s2))).isEqualTo(Integer.signum(u1.compareTo(u2)));
+  }
+
+  @ParameterizedTest
+  @ValueSource(
+      strings = {
+        "/path/file",
+        "relative/file",
+        "file:/",
+        "file:/path",
+        "file:/path/file",
+        "file:/path/file//",
+        "file:///path/file/",
+        "file:///path/file",
+        "file://///path/file",
+        "file:///path/file//",
+        "file:///path////file//",
+        "s3:/path",
+        "s3://bucket",
+        "s3://bucket/",
+      })
+  void testLocation(String input) {
```

Comment:
Proposal for testing w/ a `StorageUri.path()`:
```java
  @ParameterizedTest
  @CsvSource({
        "/path/file,/path/file",
        "relative/file,relative/file",
        "file:/,/",
        "file:/path,/path",
        "file:/path/file,/path/file",
        "file:/path/file//,/path/file/",
        "file:///path/file/,/path/file/",
        "file:///path/file,/path/file",
        "file://///path/file,/path/file",
        "file:///path/file//,/path/file/",
        "file:///path////file//,/path/file/",
        "s3:/path,/path",
        "s3://bucket,",
        "s3://bucket/,/",
      })
  void testLocation(String input, String path) {
    StorageUri s = StorageUri.of(input);
    URI u = normalizedURI(input);
    assertThat(s).extracting(StorageUri::path, StorageUri::location, StorageUri::toString).containsExactly(path, u.toString(), u.toString());
  }
```

Reply:
Added a separate method for this